week3/jsx.md

```diff
@@ -0,0 +1,292 @@
+
+```jsx
+const el = <span>JSX!</span>;
+```
+
+**Syntax extension to JavaScript**. It would remind us of a _template language_, but it's only from JavaScript.
+
+It's intended to be used by various preprocessors (transpilers) to transform these tokens into standard ECMAScript.
+
+### What is template language?
+
+Template language is the language that is **embedded** within some other master document. In the template, the document text is the most prevalent, with just a bit of code.
+
+```php
+print "Hello world with" + hj
+print "This is template." + hj + "comes here."
+```
+
+```php
+Hello world with $hj.
+This is template. $hj comes here.
+```
+
+## But JSX produces React elements
+
+So by using JSX, React can put markup and logic into one unit, called **component.**
+
+React doesn't require using JSX, but it's found helpful to use by most React users.
+
+### React without JSX
+
+Using react without JSX is convenient when you don't want to set up compilation in your build environment. (But don't you want? 😄)
+
+Each JSX element is compiled down to `React.createElement(component, props, ... children)` by Babel.
+
+So this code with JSX is compiled to the code that doesn't use JSX below:
+
+```jsx
+// With JSX
+class Hello extends React.Component {
+  render() {
+    return <div>Hello {this.props.hj}</div>;
+  }
+}
+
+ReactDOM.render(<Hello hj="Hyunjoo" />, document.getElementById("root"));
+```
+
+```jsx
+// Without JSX
+class Hello extends React.Component {
+  render() {
+    return React.createElement("div", null, `Hello ${this.props.hj}`);
+  }
+}
+
+ReactDOM.render(
+  React.createElement(Hello, { hj: "Hyunjoo" }, null),
+  document.getElementById("root")
+);
+```
+
+If we use shorthand form for `React.createElement` as below, it becomes as convenient to use React with JSX.
+
+```jsx
+// We can use shorthand of React.createElement
+const e = React.createElement;
+
+ReactDOM.render(
+	e('div', {className: 'greeting'}, 'Hello World'), document.getElementById('root')
+```
+
+Then it creates **React element object** like this:
+
+```jsx
+// Simplified
+const element = {
+  type: "div",
+  props: {
+    className: "greeting",
+    children: "Hello World",
+  },
+};
+```
+
+And React reads these objects and uses them to construct the DOM and keep it up to date.
+
+### Embedding Expressions in JSX
+
+We **put any valid JS expression** inside the curly braces in JSX. String, function, etc. We wrap JSX in parentheses to avoid the pitfalls of _Automatic Semicolon Insertion_.
+
+- Automatic Semicolon Insertion (ASI)
+
+  This rule is described in the ECMAScript specification, and it happens when the Script or Module is parsed from left to right.
+
+  Those statements are affected by the ASI.
+
+  - empty statement
+  - `var` statement
+  - expression statement
+  - `do-while` statement
+  - `continue` statement
+  - `break` statement
+  - `return` statement
+  - `throw` statement
+
+  Classic example is with the `ReturnStatement`.
+
+  ```jsx
```

Comment:
The return statements before and after the transformation are the same. It seems like a typo?